psycopg/psycopg/types/numpy.py

```diff
@@ -0,0 +1,202 @@
+        )
+
+    _int_2_dumper = NPInt2Dumper(Int2)
+    _int_4_dumper = NPInt4Dumper(Int4)
+    _int_8_dumper = NPInt8Dumper(Int8)
+    _int_numeric_dumper = NPIntNumericDumper(IntNumeric)
+
+    def get_key(self, obj: Any, format: PyFormat) -> Union[type, Tuple[type, ...]]:
+        return self.upgrade(obj, format).cls
+
+    def upgrade(self, obj: Any, format: PyFormat) -> "Dumper":
+
+        # If object is greater than the max python/postgres int size
+        # upgrade to numeric. We do the comparison here to avoid
+        # casting to
+
+        obj = int(obj)
+
+        if obj > (9_223_372_036_854_775_807):
+
+            return self._int_numeric_dumper
+
+        # For all the other cases, get the range of the value
+        # and upgrade to the appropriate dumper.
+
+        # np.uint8 bounds
+        if 0 <= obj < (16**2):
+            return self._int_2_dumper
+
+        # np.uint16 bounds
+        if (16**2) <= obj < (256**2):
+            return self._int_4_dumper
+
+        # All the values after uint16 upper limit,
+        # receive an upgrade to a Int8Dumper,
+        # so they are converted to bigint.
+
+        if (256**2) <= obj:
+            return self._int_8_dumper
+
+
+# Binary Dumpers
+
+_pack_numeric_head = cast(
+    Callable[[int, int, int, int], bytes],
+    struct.Struct("!HhHH").pack,
+)
+
+
+class NPInt2BinaryDumper(NPInt2Dumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return pack_int2(int(obj))
+
+
+class NPInt4BinaryDumper(NPInt4Dumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return pack_int4(int(obj))
+
+
+class NPInt8BinaryDumper(NPInt8Dumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return pack_int8(int(obj))
+
+
+class NPIntNumericBinaryDumper(IntNumericDumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: np.uint) -> Buffer:
+
+        # dump_int_to_numeric_binary
+
+        ndigits = obj.bit_count()
+        out = bytearray(b"\x00\x00" * (ndigits + 4))
+
+        if obj < 0:
+            sign = NUMERIC_NEG
+            obj = -obj
+        else:
+            sign = NUMERIC_POS
+
+        out[:8] = _pack_numeric_head(ndigits, ndigits - 1, sign, 0)
+        i = 8 + (ndigits - 1) * 2
+        while obj:
+            rem = obj % 10_000
+            obj //= 10_000
+            out[i : i + 2] = pack_uint2(int(rem))
+            i -= 2
+
+        return out
+
+
+class NPBooleanBinaryDumper(NPBooleanDumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return b"\x01" if obj else b"\x00"
+
+
+class NPUintBinaryDumper(NPUintDumper):
+
+    _int_2_dumper = NPInt2BinaryDumper(Int2)
+    _int_4_dumper = NPInt4BinaryDumper(Int4)
+    _int_8_dumper = NPInt8BinaryDumper(Int8)
+    _int_numeric_dumper = NPIntNumericBinaryDumper(IntNumeric)
+
+
+def register_numpy(context: AdaptContext | None = None) -> None:
```

Comment:
The  `AdaptContext | None` syntax is not supported on some of the supported Python versions (>= 3.7)